sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -190,6 +193,32 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
   }
 
   /**
+   * Wrap a DataFrameWriter action to track the query execution and time cost, then report to the
+   * user-registered callback functions.
+   *
+   * @param funcName A identifier for the method executing the query
+   * @param qe the @see `QueryExecution` object associated with the query
+   * @param outputParams The output parameters useful for query analysis
+   * @param action the function that executes the query after which the listener methods gets
+   *               called.
+   */
+  private def withAction(
+      funcName: String,
+      qe: QueryExecution,
+      outputParams: OutputParams)(action: => Unit) = {
+    try {
+      val start = System.nanoTime()
```

Comment:
`Dataset.withAction` will reset metrics of physical plans, shall we do it here? And can we create a general function for both `Dataset` and `DataFrameWriter`?